p2p/peering.go

```diff
@@ -1013,128 +1011,150 @@ func (rp *RemotePeer) Addrs(ctx context.Context) error {
 	}
 }
 
-// Block requests a block from a RemotePeer using getdata.  The same block can
-// not be requested multiple times concurrently from the same peer.
+// Block requests a block from a RemotePeer using getdata.
 func (rp *RemotePeer) Block(ctx context.Context, blockHash *chainhash.Hash) (*wire.MsgBlock, error) {
 	const opf = "remotepeer(%v).Block(%v)"
 
-	m := wire.NewMsgGetDataSizeHint(1)
-	err := m.AddInvVect(wire.NewInvVect(wire.InvTypeBlock, blockHash))
+	blocks, err := rp.Blocks(ctx, []*chainhash.Hash{blockHash})
 	if err != nil {
 		op := errors.Opf(opf, rp.raddr, blockHash)
 		return nil, errors.E(op, err)
 	}
-	c := make(chan *wire.MsgBlock, 1)
-	if !rp.addRequestedBlock(blockHash, c) {
-		op := errors.Opf(opf, rp.raddr, blockHash)
-		return nil, errors.E(op, errors.Invalid, "block is already being requested from this peer")
+
+	return blocks[0], nil
+}
+
+// requestBlocks sends a getdata wire message and waits for all the specified
+// blocks to be received. This blocks so it should be called from a goroutine.
+func (rp *RemotePeer) requestBlocks(reqs []*blockRequest) {
+	// Aux func to fulfill requests. It signals any outstanding requests of
+	// the given error and removes all from the requestedBlocks map.
+	fulfill := func(err error) {
+		rp.requestedBlocksMu.Lock()
+		for _, req := range reqs {
+			select {
+			case <-req.ready:
+				// Already fulfilled.
+			default:
+				req.err = err
+				close(req.ready)
```

Comment:
I was concerned that there may be a race here on closing the ready channel (https://play.golang.org/p/lGXEjRTkHqC) but after taking a fresh look again, I don't think this is actually possible since this only happens when the map mutex is held.  If we are going to depend on that for correctness, there should probably be a comment explaining this.

Reply:
Looking at it again, the current mutex impl prevents races but doesn't prevent a malicious peer sending a block twice in rapid succession (before the entire set of requested blocks has been processed): in that case the double close would panic the wallet.

There needs to be a check for a closed `ready` channel here as well.